### PR TITLE
feat: Add OAuth support for closecom app linking

### LINKED ISSUE #12228: Switch Close CRM integration to use OAuth
### Is your proposal related to a problem?

When the [Close CRM app](https://app.cal.com/apps/closecom) was created, Close only supported Basic Auth via API Key.

Now Close supports OAuth:
https://developer.close.com/topics/authentication-oauth2/

### Describe the solution you'd like

It would be great if the integration could be converted so that at least new users would be prompted to connect via OAuth instead.

### Additional context

Beyond ease of set up and better security for users, there may be other benefits that Close gives toward OAuth apps. For example it also means Cal could take advantage of Close's new [Meeting Integration Notes feature](https://developer.close.com/resources/activities/meeting/#create-or-update-third-party-meeting-integration) which requires an OAuth connection. Similarly, Close's new [Scheduling Link feature](https://developer.close.com/resources/scheduling-links/) can also be more sophisticated once there's an OAuth connection.

Related to:
https://github.com/calcom/cal.com/blob/d1d50b0d9131e650eab2a88396e130dfb2067da4/packages/app-store/closecom/lib/CalendarService.ts#L34-L38

Get in touch with phil@close.com or support@close.com if you want to collaborate on a deeper integration!

### PR DESCRIPTION
**Pull request by asadath1395**
## What does this PR do?

Add OAuth support for closecom app linking

- Fixes #12228 

https://www.loom.com/share/9f0ab33343d74f9799abe518556365c2?sid=fb2663bf-9ce1-495c-a00a-cb23ea020825

## Mandatory Tasks (DO NOT REMOVE)

- [x] I have self-reviewed the code (A decent size PR without self-review might be rejected).
- [x] I have updated the developer docs in /docs if this PR makes changes that would require a [documentation change](https://cal.com/docs). If N/A, write N/A here and check the checkbox.
- [x] I confirm automated tests are in place that prove my fix is effective or that my feature works.

## How should this be tested?

- Create an account in [Close](https://www.close.com/)
- Create an OAuth app 
- Make note of the `client_id` and `client_secret` and add those as environment variables in `.env`
- Install the app

### CODE REVIEW

**Review by Praashh — CHANGES_REQUESTED** [review]
type checks are failing, can you please fix ?

**Comment by asadath1395** [comment]
@Praashh Fixed tests and added loom video too

**Review by Praashh — APPROVED** [review]
LGTM!

**Comment by anikdhabal** [comment]
What about the current users who are using the legacy one right now?

File: packages/app-store/closecom/lib/CrmService.ts [hunk 1]
```diff
@@ -55,17 +66,30 @@ export default class CloseComCRMService implements CRM {
...
-    if (parsedCredentialKey.success) {
-      decrypted = symmetricDecrypt(parsedCredentialKey.data.encrypted, CALENDSO_ENCRYPTION_KEY);
+    if (!parsedKey.success) {
+      throw new Error(
+        `Invalid credentials for userId ${credential.userId} and appId ${credential.appId}: ${parsedKey.error}`
+      );
+    }
+
+    // Initialize CloseCom client based on credential type
+    if (parsedKey.data.encrypted) {
```
Comment (asadath1395) [inline packages/app-store/closecom/lib/CrmService.ts:78]: Keep support for legacy API key

File: packages/lib/CloseCom.ts [hunk 2]
```diff
@@ -186,21 +185,136 @@ export type CloseComCustomFieldCreateResponse = {
...
-    this.apiKey = providedApiKey || environmentApiKey;
+    this.api_key = apiKey;
+    this.refresh_token = options?.refresh_token;
+    this.expires_at = options?.expires_at;
+    this.isOAuth = options?.isOAuth || false;
+    this.userId = options?.userId;
+  }
+
+  private async getHeaders() {
+    if (this.isOAuth) {
```
Comment (asadath1395) [inline packages/lib/CloseCom.ts:221]: Support both OAuth and API key based authentication

**Comment by asadath1395** [comment]
> @asadath1395 What about the current users who are using the legacy one right now?

@anikdhabal Its already handled in code right now, we are still supporting old API key based approach along with new OAuth approach. User's who install/reinstall the app will use new OAuth, old users will still be using API key based authentication

**Comment by keithwillcode** [comment]
Would like @emrysal or @zomars to review this before merging

**Review by emrysal — APPROVED** [review]
Looking good to me 👏  - nice work.

**Pull request merged**